### PR TITLE
[ENG-1049] Gracefully Handle ORCiD Login Exceptions

### PR DESCRIPTION
**Pull request by cslzchen**
## Ticket

https://openscience.atlassian.net/browse/ENG-1049

## Purpose

Whenever something went wrong in `pac4j` (a library we use), it crashes CAS. This PR  fixes this issue by rewiring the login web flow so that CAS can handle exceptions from ORCiD login (and other delegated login including CAS) gracefully.

### Major Change

* New exceptions for delegated login failures
* New views (pages) for the new exceptions
* The auth handlers throw the generic delegated login exception
* The client action throws the client specific delegated login exception
* The client action catches the exceptions (including ones thrown by itself) and returns a new event that redirects the login web flow to the main CAS auth exception handler.
* The main exception handler makes takes care of the rest and redirect users to the new exception pages accordingly

### Minor Tweaks

* Use access token (via `Authorization` header instead of query parameter) when making ORCiD public API calls.
* Replaced deprecated default scope with `/authenticate`. This has no effect since the default scope will be overwritten by custom server config, which happens to be `/authenticate` anyway.
* Revoked prefix "www." from the authorize url.
* Refactored the code style for overlayed classes touched in this PR so that functionality changes are in their own commits and more readable.

### Side Effect

Handles exceptions for all `pac4j`-delegated login gracefully, including institution login via the `CasClient`.

## Dev / QA Notes

* Dev Regression Test

### New Exception Pages

It is very hard to trigger errors (e.g. network issue, permission denied, etc.) during delegated login. Here is how to manually trigger the error to see the new pages.

* ORCiD Client failure due to invalid or expired authorization code:

	https://accounts.staging.osf.io/login?client_name=OrcidClient&code=INVALID_OR_EXPIRED_CODE&service=https%3A%2F%2Fstaging.osf.io%2Flogin%2F%3Fnext%3Dhttps%253A%252F%252Fstaging.osf.io%252F

* CAS Client failure due to invalid or expired service ticket:

	https://accounts.staging.osf.io/login?client_name=callutheran&ticket=INVALID_OR_EXPIRED_TICKET&service=https%3A%2F%2Fstaging.osf.io%2Flogin%2F%3Fnext%3Dhttps%253A%252F%252Fstaging.osf.io%252F

* General authentication delegation failures. For example, invalid client:

	https://accounts.staging.osf.io/login?client_name=INVALID_CLIENT_NAME&service=https%3A%2F%2Fstaging.osf.io%2Flogin%2F%3Fnext%3Dhttps%253A%252F%252Fstaging.osf.io%252F
## Dev-Ops Notes

No

### CODE REVIEW

File: cas-server-support-osf/src/main/java/org/pac4j/oauth/client/OrcidClient.java [hunk 1]
```diff
@@ -16,33 +16,33 @@
...
+import org.scribe.model.ProxyOAuthRequest;
+import org.scribe.model.Response;
 import org.scribe.model.SignatureType;
 import org.scribe.model.Token;
-import org.scribe.oauth.ProxyOAuth20ServiceImpl;
 import org.scribe.tokens.OrcidToken;
-import org.scribe.model.ProxyOAuthRequest;
-import org.scribe.model.Response;
-
+import org.scribe.oauth.ProxyOAuth20ServiceImpl;
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/org/pac4j/oauth/client/OrcidClient.java:35]: Wrong order, please fix.

File: cas-server-support-osf/src/main/java/org/pac4j/oauth/client/OrcidClient.java [hunk 2]
```diff
@@ -239,6 +240,9 @@ protected String sendRequestForData(
         logger.debug("accessToken : {} / dataUrl : {}", accessToken, dataUrl);
         final long t0 = System.currentTimeMillis();
         final ProxyOAuthRequest request = createProxyRequest(dataUrl);
+        if (accessToken != null) {
+            request.addHeader("Authorization", "Bearer " + accessToken.getToken());
+        }
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/org/pac4j/oauth/client/OrcidClient.java:256]: The following piece of code is in the [overridden method](https://github.com/pac4j/pac4j/blob/1.7.x/pac4j-oauth/src/main/java/org/pac4j/oauth/client/BaseOAuthClient.java#L223-L227). For ORCiD, the `signRequest()` just appends the access token as a query parameter, which is very dangerous. In addition, the `isTokenAsHeader()` is always true for the [OrcidClient](https://github.com/pac4j/pac4j/blob/1.7.x/pac4j-oauth/src/main/java/org/pac4j/oauth/client/OrcidClient.java#L47-L55).
```java
...
final ProxyOAuthRequest request = createProxyRequest(dataUrl);
this.service.signRequest(accessToken, request);
// Let the client to decide if the token should be in header
if (this.isTokenAsHeader()) {
        request.addHeader("Authorization", "Bearer " + accessToken.getToken());
}
final Response response = request.send();
...
```
It's is unclear what issue made me remove the header along with the unsafe token query initially. It could be a mistake or the `/authenticate` scope with public API wouldn't accept access token at that time. After revisiting both the overriding and the overridden method, I decided to put the header back.

File: cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkAuthenticationExceptionHandler.java [hunk 3]
```diff
@@ -23,6 +23,9 @@
 import java.util.List;
 import java.util.Set;
 
+import io.cos.cas.authentication.exceptions.CasClientLoginException;
+import io.cos.cas.authentication.exceptions.DelegatedLoginException;
+import io.cos.cas.authentication.exceptions.OrcidClientLoginException;
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkAuthenticationExceptionHandler.java:28]: I decided to put new exceptions in one place under the `exceptions` package. Existing ones will be moved in another refactor PR.

File: cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/AbstractClientAuthenticationHandler.java [hunk 4]
```diff
@@ -126,12 +147,11 @@ protected HandlerResult doAuthentication(
      * @param profile the retrieved user profile
      * @return the built handler result
      * @throws GeneralSecurityException on authentication failure
-     * @throws PreventedException on the indeterminate case when authentication is prevented
      */
     protected abstract HandlerResult createResult(
             final ClientCredential credentials,
             final UserProfile profile
-    ) throws GeneralSecurityException, PreventedException;
+    ) throws GeneralSecurityException;
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/AbstractClientAuthenticationHandler.java:148]: In Java, it is OK for implementation classes not to throw all exceptions declared in the interface.

File: cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/ClientAuthenticationHandler.java [hunk 5]
```diff
@@ -63,34 +62,28 @@ public ClientAuthenticationHandler(final Clients theClients) {
     protected HandlerResult createResult(
             final ClientCredential credentials,
             final UserProfile profile
-    ) throws GeneralSecurityException, PreventedException {
+    ) throws GeneralSecurityException {
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/ClientAuthenticationHandler.java:65]: Ditto, removed redundant exceptions.

File: cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/ClientAuthenticationHandler.java [hunk 6]
```diff
@@ -63,34 +62,28 @@ public ClientAuthenticationHandler(final Clients theClients) {
...
         }
-        throw new FailedLoginException("No identifier found for this user profile: " + profile);
+        throw new DelegatedLoginException("No identifier found for this user profile: " + profile);
     }
 
-    /**
-     * Check if type ID is used.
-     *
-     * @return the type ID flag
-     */
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/ClientAuthenticationHandler.java]: In java, getter and setters don't need JavaDoc. In addition, for `boolean` properties, getter is prefixed with `is` instead of `get`.

File: cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/ClientAuthenticationHandler.java [hunk 7]
```diff
@@ -63,34 +62,28 @@ public ClientAuthenticationHandler(final Clients theClients) {
...
     public boolean isTypedIdUsed() {
         return typedIdUsed;
     }
 
-    /**
-     * Toggle the type ID flag on and off.
-     *
-     * @param typedIdUsed a boolean value to set the type ID flag
-     */
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/ClientAuthenticationHandler.java]: Diddo, no need for JavaDoc

File: cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java [hunk 8]
```diff
@@ -177,20 +197,30 @@ protected Event doExecute(final RequestContext context) throws Exception {
...
+            // If any exception occurs during the client authentication process, return the specialized event which
+            // will allow the CAS authentication exception handler to handle the exception gracefully.
+            final Map<String, Class<? extends Exception>> failures = new LinkedHashMap<>();
+            failures.put(e.getClass().getSimpleName(), e.getClass());
+            return getEventFactorySupport().event(
+                    this,
+                    AUTHENTICATION_FAILURE,
+                    new LocalAttributeMap<>("error", new AuthenticationException(failures))
+            );
+        }
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java]: Turn all exceptions into an event `AUTHENTICATION_FAILURE` for the login web flow to redirect to the primary CAS authentication handler. The exception detail is stored in the event attributes.

File: cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/AbstractClientAuthenticationHandler.java [hunk 9]
```diff
@@ -119,17 +115,12 @@ protected HandlerResult doAuthentication(final Credential credential) throws Gen
...
         final UserProfile userProfile;
         try {
             userProfile = client.getUserProfile(credentials, webContext);
             logger.debug("userProfile : {}", userProfile);
         } catch (final TechnicalException e) {
-            if (OrcidClient.class.getSimpleName().equals(client.getName())) {
-                throw new OrcidClientLoginException(e.getMessage());
-            } else if (CasClient.class.getSimpleName().equals(client.getName())) {
-                throw new CasClientLoginException(e.getMessage());
-            }
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/AbstractClientAuthenticationHandler.java]: Exception specialization removed from the client auth handler.

File: cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java [hunk 10]
```diff
@@ -199,10 +206,14 @@ protected Event doExecute(final RequestContext context) throws Exception {
...
-                WebUtils.putTicketGrantingTicketInScopes(context, tgt);
-                return success();
+                try {
+                    final TicketGrantingTicket tgt = this.centralAuthenticationService
+                            .createTicketGrantingTicket(new ClientCredential(credentials));
+                    WebUtils.putTicketGrantingTicketInScopes(context, tgt);
+                    return success();
+                } catch (final Exception e) {
+                    throw createClientSpecificDelegatedLoginException(client.getClass().getSimpleName(), e);
+                }
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java]: Exception specialization added to the client action.

File: cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java [hunk 11]
```diff
@@ -185,6 +190,8 @@ protected Event doExecute(final RequestContext context) throws Exception {
                 final ExternalContext externalContext = ExternalContextHolder.getExternalContext();
                 externalContext.recordResponseComplete();
                 return new Event(this, "stop");
+            } catch (final CredentialsException e) {
+                throw createClientSpecificDelegatedLoginException(client.getClass().getSimpleName(), e);
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java]: Added the missing credential exception which should have been caught and specialized.

Fyi, an example of credential exception for OAuth 2.0 clients (e.g. ORCiD) can be expired/invalid/missing authorization code. another for CAS client (e.g. institutions) can be expired/invalid/missing service ticket.

File: cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/delegatedLoginCasErrorView.jsp [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    Unless required by applicable law or agreed to in writing,
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+--%>
+
+<%-- Login exception page: invalid or unexpected user status --%>
```
Comment (cslzchen) [inline cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/delegatedLoginCasErrorView.jsp]: Shame!!! Copy-n-paste but comments are left. Fix it.

File: cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/delegatedLoginGenericErrorView.jsp [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    Unless required by applicable law or agreed to in writing,
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+--%>
+
+<%-- Login exception page: invalid or unexpected user status --%>
```
Comment (cslzchen) [inline cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/delegatedLoginGenericErrorView.jsp]: Ditto, fix comments

File: cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/delegatedLoginOrcidErrorView.jsp [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+    Unless required by applicable law or agreed to in writing,
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+--%>
+
+<%-- Login exception page: invalid or unexpected user status --%>
```
Comment (cslzchen) [inline cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/delegatedLoginOrcidErrorView.jsp]: Ditto, fix comments

**Review by cslzchen — COMMENTED** [review]
Please refer to the comments for minor patches.

File: cas-server-support-osf/src/main/java/org/pac4j/oauth/client/OrcidClient.java [hunk 15]
```diff
@@ -16,39 +16,39 @@
...
  * @author Michael Haselton
  * @author Longze Chen
- * @see org.pac4j.oauth.profile.orcid.OrcidProfile
  * @since 1.7.1
  */
 public class OrcidClient extends BaseOAuth20Client<OrcidProfile> {
 
     /** The default scope. */
-    protected static final String DEFAULT_SCOPE = "/orcid-profile/read-limited";
+    protected static final String DEFAULT_SCOPE = "/authenticate";
```
Comment (mfraezz) [inline cas-server-support-osf/src/main/java/org/pac4j/oauth/client/OrcidClient.java:62]: Did orcid change the necessary scope, or did we get it wrong? A link to their scope docs might be helpful here

File: cas-server-support-osf/src/main/java/org/pac4j/oauth/client/OrcidClient.java [hunk 16]
```diff
@@ -167,20 +168,26 @@ public void setMember(final Boolean member) {
...
     protected OrcidProfile extractUserProfile(final String body) {
+
         final OrcidProfile profile = new OrcidProfile();
         profile.setId(XmlHelper.get(body, OrcidAttributesDefinition.ORCID));
+
         for(final String attribute : OAuthAttributesDefinitions.orcidDefinition.getAllAttributes()) {
             final String value = XmlHelper.get(body, attribute);
             switch (attribute) {
                 case OrcidAttributesDefinition.NORMALIZED_FAMILY_NAME:
-                    break;
```
Comment (mfraezz) [inline cas-server-support-osf/src/main/java/org/pac4j/oauth/client/OrcidClient.java:183]: Why was this removed?

File: cas-server-webapp/src/main/resources/messages.properties [hunk 17]
```diff
@@ -166,7 +183,7 @@ screen.institution.login.select.error.message=You must select an institution.
 
 # Sign in through OSF
 screen.osf.login.message=Sign in with your OSF account to continue
-screen.osf.login.message.error=Oops! Something went wrong
+screen.osf.login.message.error=Oops! Something went wrong ...
```
Comment (mfraezz) [inline cas-server-webapp/src/main/resources/messages.properties:186]: Do we do this anywhere else?

File: cas-server-support-osf/src/main/java/org/pac4j/oauth/client/OrcidClient.java (hunk 16 above)
Comment (cslzchen) [reply cas-server-support-osf/src/main/java/org/pac4j/oauth/client/OrcidClient.java:183]: I had the same question when I saw the style warning for this `break;` statement. I went on and searched. It turns out that for Java `switch` statement, if the same action (here just the `break;`) is used for multiple cases, I should group the cases and just write the action once.

File: cas-server-webapp/src/main/resources/messages.properties (hunk 17 above)
Comment (cslzchen) [reply cas-server-webapp/src/main/resources/messages.properties:186]: As far as CAS is concerned, this is the message displayed to the users for all CAS failures. I had planned to add the `...` a while ago but never did. (It was in the strangled huge CAS API / Account Manager PR lol) So I just slipped it in here.

File: cas-server-support-osf/src/main/java/org/pac4j/oauth/client/OrcidClient.java (hunk 15 above)
Comment (cslzchen) [reply cas-server-support-osf/src/main/java/org/pac4j/oauth/client/OrcidClient.java:62]: This is just the default one provided by `pac4j` and it is no longer a valid scope. It will always be overwritten by the settings in `cas.properties` (which is `/authenticate`). I like the idea of adding the scope docs: https://members.orcid.org/api/oauth/orcid-scopes.

File: cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/AbstractClientAuthenticationHandler.java [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+        final String clientName = credentials.getClientName();
+        logger.debug("clientName : {}", clientName);
+
+        // Attempt to load the client from defined clients
+        final Client<Credentials, UserProfile> client;
+        try {
+            client = this.clients.findClient(clientName);
+            logger.debug("client : {}", client);
+        } catch (final TechnicalException e) {
+            throw new DelegatedLoginException("Invalid client: " + clientName);
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/AbstractClientAuthenticationHandler.java:109]: Please add an error-level log when exception happens so that we can pinpoint the issue when login fails.

File: cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/AbstractClientAuthenticationHandler.java [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+        final HttpServletResponse response = (HttpServletResponse) servletExternalContext.getNativeResponse();
+        final WebContext webContext = new J2EContext(request, response);
+
+        // Attempt to retrieve the user profile
+        final UserProfile userProfile;
+        try {
+            userProfile = client.getUserProfile(credentials, webContext);
+            logger.debug("userProfile : {}", userProfile);
+        } catch (final TechnicalException e) {
+            throw new DelegatedLoginException(e.getMessage());
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/AbstractClientAuthenticationHandler.java:126]: Similarly, add an error-level log.

File: cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/AbstractClientAuthenticationHandler.java [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+        }
+
+        // If successful, create and return a handler result; otherwise, throw a delegated login exception.
+        if (userProfile != null) {
+            return createResult(clientCredentials, userProfile);
+        }
+        throw new DelegatedLoginException("Provider did not produce a user profile for: " + clientCredentials);
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/AbstractClientAuthenticationHandler.java:134]: Ditto, log! log! log!

File: cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/ClientAuthenticationHandler.java [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+        if (StringUtils.isNotBlank(id)) {
+            try {
+                credentials.setUserProfile(profile);
+                credentials.setTypedIdUsed(typedIdUsed);
+                return new DefaultHandlerResult(
+                        this,
+                        new BasicCredentialMetaData(credentials),
+                        this.principalFactory.createPrincipal(id, profile.getAttributes()));
+            } catch (final Exception e) {
+                throw new DelegatedLoginException(e.getMessage());
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/ClientAuthenticationHandler.java:78]: Log

File: cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/ClientAuthenticationHandler.java [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+            }
+        }
+        throw new DelegatedLoginException("No identifier found for this user profile: " + profile);
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/ClientAuthenticationHandler.java:82]: Log

File: cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java [hunk 23]
```diff
@@ -0,0 +1,321 @@
...
+
+            // If `client_name` found in the request parameter, this is an authentication. Verify the client and the
+            // protocol; retrieve credentials and session parameters; and attempt to authenticate.
+            // 1. Retrieve the client
+            final BaseClient<Credentials, CommonProfile> client;
+            try {
+                client = (BaseClient<Credentials, CommonProfile>) this.clients.findClient(clientName);
+                logger.debug("client: {}", client);
+            } catch (final TechnicalException e) {
+                throw new DelegatedLoginException("Invalid client: " + clientName);
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java:175]: Log

File: cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java [hunk 24]
```diff
@@ -0,0 +1,321 @@
...
+                logger.debug("credentials: {}", credentials);
+            } catch (final RequiresHttpAction e) {
+                // Requires HTTP action: client redirection
+                logger.debug("requires http action: {}", e.toString());
+                response.flushBuffer();
+                final ExternalContext externalContext = ExternalContextHolder.getExternalContext();
+                externalContext.recordResponseComplete();
+                return new Event(this, "stop");
+            } catch (final CredentialsException e) {
+                throw createClientSpecificDelegatedLoginException(client.getClass().getSimpleName(), e);
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java:197]: Log

File: cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java [hunk 25]
```diff
@@ -0,0 +1,321 @@
...
+
+            if (credentials != null) {
+                // 5. Attempt to authenticate if the credential is not null.
+                try {
+                    final TicketGrantingTicket tgt = this.centralAuthenticationService
+                            .createTicketGrantingTicket(new ClientCredential(credentials));
+                    WebUtils.putTicketGrantingTicketInScopes(context, tgt);
+                    return success();
+                } catch (final Exception e) {
+                    throw createClientSpecificDelegatedLoginException(client.getClass().getSimpleName(), e);
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java:219]: Log

File: cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java [hunk 26]
```diff
@@ -0,0 +1,321 @@
...
+        } catch (final Exception e) {
+            // If any exception occurs during the client authentication process, return the specialized event which
+            // will allow the CAS authentication exception handler to handle the exception gracefully.
+            final Map<String, Class<? extends Exception>> failures = new LinkedHashMap<>();
+            failures.put(e.getClass().getSimpleName(), e.getClass());
+            return getEventFactorySupport().event(
+                    this,
+                    AUTHENTICATION_FAILURE,
+                    new LocalAttributeMap<>("error", new AuthenticationException(failures))
+            );
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java]: Add a error-level log in case unexpected errors are caught.

**Review by cslzchen — COMMENTED** [review]
Please add error-level logs, see comments for details.

**Review by cslzchen — COMMENTED** [review]
Looks good and will be merged into [`feature/improve-delegated-login`](https://github.com/CenterForOpenScience/cas-overlay/tree/feature/improve-delegated-login).
